### PR TITLE
filter patients on role

### PR DESCRIPTION
**Pull request by pbugni**
Now filters out patients IF the authorized user has the `org_admin` or `org_staff` role.

### CODE REVIEW

**Comment by pbugni** [comment]
Second guessing decision to use hyphens in role names.  Think I'll convert that to `org_admin` and `org_staff`

File: map/migrations/versions/purge_bogus_consents.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+def upgrade():
+    """Consents are now updated in place.  Delete obsolete"""
+    # Consents are between a Patient and an Organization, then categorized
+    # by a list of Code/Systems.  For each f(Patient, Org, Code) keep only
+    # the one with the most recent `startDate`
+
+    # working with our known Organizations, get the respective list of consents
+    # for each
+    consent_ids_to_delete = []
+    for org_id in (1463, 1466, 1467, 1467, 1737):
```
Comment (hannahburkhardt) [inline map/migrations/versions/purge_bogus_consents.py]:
```suggestion
    for org_id in (1463, 1464, 1465, 1466, 1467, 1737):
```

File: map/authz/authorizeduser.py [hunk 2]
```diff
@@ -147,21 +147,40 @@ def _kc_ident_in_resource(self):
...
     def read(self):
-        """Only owning patient may read"""
+        """User's role determines read access"""
+        # Admins get carte blanche
         if 'admin' in self.user.roles:
-            return True
-        if not self._kc_ident_in_resource():
+            return self.resource
+
+        # Org admin and staff can only view patients with consents
```
Comment (ivan-c) [inline map/authz/authorizeduser.py:166]: I like these plain-text rule explanations. Feels like we could codify them into policy at some later date

File: map/authz/authorizeduser.py [hunk 3]
```diff
@@ -260,22 +282,90 @@ def check(self, verb, fhir):
...
+        """Lookup all users with consent on given org"""
+        if not org_id:
+            return set()
+
+        # shouldn't need to round trip twice
+        if hasattr(self, '_consented_users'):
+            return self._consented_users
+
+        self._consented_users = set()
+        result, status = HapiRequest.find_bundle('Consent', search_dict={
```
Comment (ivan-c) [inline map/authz/authorizeduser.py:324]: might be more readable to break this down into a key per line for `search_dict`

File: map/authz/authorizeduser.py [hunk 4]
```diff
@@ -260,22 +282,90 @@ def check(self, verb, fhir):
...
+            'organization': '/'.join(("Organization", str(org_id))),
+            '_include': "Consent.patient", '_count': 1000})
+        bundle = Bundle(result)
+        for i in bundle.resources():
+            if i['resourceType'] == 'Consent' and i[
```
Comment (ivan-c) [inline map/authz/authorizeduser.py]: prefer to linebreak after `and`, to prevent splitting the array index lookup

File: map/fhir/bundle.py [hunk 5]
```diff
@@ -15,3 +18,25 @@ def resources(self):
...
         for entry in self.bundle['entry']:
             yield entry["resource"]
+
+    def remove_entries(self, ids):
+        """remove given entries from the contained bundle
+
+        :param ids: iterable of id values to remove from bundle
+        :raises ValueError: if matching entries not found
+        """
+        count_b4 = self.__len__()
```
Comment (ivan-c) [inline map/fhir/bundle.py]: why doesn't `len()` work here?...

File: tests/test_authz.py [hunk 6]
```diff
@@ -28,11 +32,26 @@ def admin_jwt():
...
 @fixture
 def prefix(app):
     return app.config.get('API_PREFIX')
 
 
+@fixture
+def consented_patient_bundle(request):
+    data_dir, _ = os.path.splitext(request.module.__file__)
+    with open(os.path.join(
+            data_dir, "consented_patient.json"), 'r') as json_file:
```
Comment (ivan-c) [inline tests/test_authz.py:50]: going to steal this fixture for the PDMP XML response :D

**Review by ivan-c — APPROVED** [review]
Looks good! Had some questions but no blockers, merge when ready

File: map/fhir/bundle.py (hunk 5 above)
Comment (pbugni) [reply map/fhir/bundle.py]: it does - thanks.

**Pull request merged**